lib/resumption/managedconn.go

```diff
@@ -0,0 +1,449 @@
+		return w.data[left:], w.data[:left]
+	}
+
+	left, right := w.bounds()
+
+	if left >= right {
+		return w.data[right:left], nil
+	}
+	return w.data[right:], w.data[:left]
+}
+
+// reserve ensures that the buffer has a given amount of free space,
+// reallocating its internal buffer as needed. After reserve(n), the two slices
+// returned by free have total length at least n.
+func (w *buffer) reserve(n uint64) {
+	n += w.len()
+	if n <= len64(w.data) {
+		return
+	}
+
+	newCapacity := max(len64(w.data)*2, initialBufferSize)
+	for n > newCapacity {
+		newCapacity *= 2
+	}
+
+	d1, d2 := w.buffered()
+	w.data = make([]byte, newCapacity)
+	w.end = w.start
+
+	// this is less efficient than copying the data manually, but almost all
+	// uses of buffer will eventually hit a maximum buffer size anyway
+	w.append(d1)
+	w.append(d2)
+}
+
+// append copies the slice to the tail of the buffer, resizing it if necessary.
+// Writing to the slices returned by free() and appending them in order will not
+// result in any memory copy (if the buffer hasn't been reallocated).
+func (w *buffer) append(b []byte) {
+	w.reserve(len64(b))
+	f1, f2 := w.free()
+	// after reserve(n), len(f1)+len(f2) >= n, so this is guaranteed to work
+	copy(f2, b[copy(f1, b):])
+	w.end += len64(b)
+}
+
+// write copies the slice to the tail of the buffer like in append, but only up
+// to the total buffer size specified by max. Returns the count of bytes copied
+// in, which is always not greater than len(b) and (max-w.len()).
+func (w *buffer) write(b []byte, max uint64) uint64 {
+	if w.len() >= max {
+		return 0
+	}
+
+	s := min(max-w.len(), len64(b))
+	w.append(b[:s])
+	return s
+}
+
+// advance will discard bytes from the head of the buffer, advancing its start
+// position. Advancing past the end causes the end to be pushed forwards as
+// well, such that an empty buffer advanced by n ends up with start = end = n.
+func (w *buffer) advance(n uint64) {
+	w.start += n
+	if w.start > w.end {
+		w.end = w.start
+	}
+}
+
+// read will attempt to fill the slice with as much data from the buffer,
+// advancing the start position of the buffer to match. Returns the amount of
+// bytes copied in the slice.
+func (w *buffer) read(b []byte) int {
+	d1, d2 := w.buffered()
+	n := copy(b, d1)
+	n += copy(b[n:], d2)
+	w.advance(uint64(n))
+	return n
+}
+
+// deadline holds the state necessary to handle [net.Conn]-like deadlines.
+// Should be paired with a [sync.Cond], whose lock protects access to the data
+// inside the deadline, and that will get awakened if and when the timeout is
+// reached.
+type deadline struct {
+	// deadline should not be moved or copied
+	_ [0]sync.Mutex
+
+	// timeout is true if we're past the deadline.
+	timeout bool
+
+	// stopped is set if timer is non-nil but it's stopped and ready for reuse.
+	stopped bool
+	// timer, if set, is a [time.AfterFunc] timer that sets timeout after
+	// reaching the deadline. Initialized on first use.
+	timer clockwork.Timer
```

Comment:
I'm not getting better alignment but I'm getting fewer bytes that the gc needs to check (24 vs 16). Thank you for making me aware of this additional tidbit, now I'll spend time obsessing over my struct layout even more. 😔

(`managedConn` itself is going to go from 224 to 184 pointer bytes after some rearrangement)

Reply:
Yep, I used that one; I wasn't sure of the CLI usage, but it can be added to squigglies in vscode with
```json
    "gopls": {
        "ui.diagnostic.analyses": {
            "fieldalignment": false,
        }
    },
```